processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java

```diff
@@ -0,0 +1,457 @@
+        {
+          final int offset = buffer.getInt(tableArenaSize + curr * Ints.BYTES);
+          final Entry<KeyType> entry = bucketEntryForOffset(offset);
+          curr++;
+
+          return entry;
+        }
+
+        @Override
+        public void remove()
+        {
+          throw new UnsupportedOperationException();
+        }
+      };
+    }
+  }
+
+  @Override
+  public void close()
+  {
+    for (BufferAggregator aggregator : aggregators) {
+      aggregator.close();
+    }
+  }
+
+  private boolean isUsed(final int bucket)
+  {
+    return (tableBuffer.get(bucket * bucketSize) & 0x80) == 0x80;
+  }
+
+  private Entry<KeyType> bucketEntryForOffset(final int bucketOffset)
+  {
+    final KeyType key = keySerde.fromByteBuffer(tableBuffer, bucketOffset + HASH_SIZE);
+    final Object[] values = new Object[aggregators.length];
+    for (int i = 0; i < aggregators.length; i++) {
+      values[i] = aggregators[i].get(tableBuffer, bucketOffset + aggregatorOffsets[i]);
+    }
+
+    return new Entry<>(key, values);
+  }
+
+  private void growIfPossible()
+  {
+    final int newBuckets = buckets * 2;
+    final int newMaxSize = maxSizeForBuckets(newBuckets);
+
+    if (newBuckets > (tableArenaSize - tableStart) / bucketSize) {
+      // Not enough space left to grow
+      return;
+    }
+
+    final int newTableStart = tableStart + tableBuffer.limit();
+
+    ByteBuffer newTableBuffer = buffer.duplicate();
+    newTableBuffer.position(newTableStart);
+    newTableBuffer.limit(newTableBuffer.position() + newBuckets * bucketSize);
+    newTableBuffer = newTableBuffer.slice();
+
+    int newSize = 0;
+
+    // Clear used bits of new table
+    for (int i = 0; i < newBuckets; i++) {
+      newTableBuffer.put(i * bucketSize, (byte) 0);
+    }
+
+    // Loop over old buckets and copy to new table
+    for (int oldBucket = 0; oldBucket < buckets; oldBucket++) {
+      if (isUsed(oldBucket)) {
+        final ByteBuffer entryBuffer = tableBuffer.duplicate();
+        entryBuffer.position(oldBucket * bucketSize);
+        entryBuffer.limit((oldBucket + 1) * bucketSize);
+
+        final ByteBuffer keyBuffer = entryBuffer.duplicate();
+        keyBuffer.position(entryBuffer.position() + HASH_SIZE);
+        keyBuffer.limit(keyBuffer.position() + keySize);
+
+        final int keyHash = entryBuffer.getInt(entryBuffer.position()) & 0x7fffffff;
+
+        final int newBucket = findBucket(newTableBuffer, newBuckets, bucketSize, true, keyBuffer, keySize, keyHash);
+        if (newBucket < 0) {
+          throw new ISE("WTF?! Couldn't find a bucket while resizing?!");
+        }
+        newTableBuffer.position(newBucket * bucketSize);
+        newTableBuffer.put(entryBuffer);
+
+        buffer.putInt(tableArenaSize + newSize * Ints.BYTES, newBucket * bucketSize);
+        newSize++;
+      }
+    }
+
+    buckets = newBuckets;
+    maxSize = newMaxSize;
+    tableBuffer = newTableBuffer;
+    tableStart = newTableStart;
```

Comment:
am i correct in assuming that , on each "grow" old tableBuffer space is not reused afterwards ?